File: webllm/ast_parser.mjs
// ast_parser.mjs
// XML boolean AST parser for sovereign routing expressions
// Handles: <route>, <and>, <or>, <not>, <leaf op="" weight="" valid=""/>
// No external XML parser — pure regex-based recursive descent

const BOOLEAN_TAGS = new Set(['and', 'or', 'not', 'route']);
const VALID_OPS = ['ADD', 'MUL', 'XOR', 'LOOP', 'MEMCPY', 'MEMSET', 'STRCMP', 'HELLO'];

// Parse a <leaf .../> element attributes
function parseLeafAttrs(tag) {
  const opMatch = tag.match(/op="([^"]+)"/);
  const weightMatch = tag.match(/weight="([^"]+)"/);
  const validMatch = tag.match(/valid="([^"]+)"/);
  return {
    op: opMatch ? opMatch[1].toUpperCase() : null,
    weight: weightMatch ? parseFloat(weightMatch[1]) : 0.05,
    valid: validMatch ? validMatch[1] === 'true' : false,
  };
}

// Find the next open tag <tagName with valid boundary char (>, space, /, newline, tab)
function findNextOpenTag(xml, tagName, start) {
  let i = start;
  const needle = '<' + tagName;
  while (i < xml.length) {
    const idx = xml.indexOf(needle, i);
    if (idx === -1) return -1;
    const charAfter = xml[idx + 1 + tagName.length];
    if (
      charAfter === '>' ||
      charAfter === ' ' ||
      charAfter === '\t' ||
      charAfter === '\n' ||
      charAfter === '/' ||
      charAfter === undefined
    ) {
      return idx;
    }
    i = idx + 1;
  }
  return -1;
}

// Extract direct children tags from an XML string (non-recursive, one level deep)
// Returns array of { tag, content, selfClosing, raw }
function extractChildren(xml) {
  const children = [];
  let i = 0;

  while (i < xml.length) {
    // Skip to next '<'
    while (i < xml.length && xml[i] !== '<') i++;
    if (i >= xml.length) break;

    // Skip HTML comments
    if (xml.slice(i, i + 4) === '<!--') {
      const end = xml.indexOf('-->', i);
      i = end === -1 ? xml.length : end + 3;
      continue;
    }

    // Skip closing tags
    if (xml[i + 1] === '/') {
      const end = xml.indexOf('>', i);
      i = end === -1 ? xml.length : end + 1;
      continue;
    }

    // Skip processing instructions
    if (xml[i + 1] === '?') {
      const end = xml.indexOf('?>', i);
      i = end === -1 ? xml.length : end + 2;
      continue;
    }

    const tagStart = i;
    const gtIdx = xml.indexOf('>', i);
    if (gtIdx === -1) { i++; continue; }

    // Everything between < and >
    const tagInner = xml.slice(i + 1, gtIdx);

    // Self-closing: ends with /
    if (tagInner.endsWith('/')) {
      const tagName = tagInner.split(/[\s/]/)[0];
      children.push({
        tag: tagName,
        content: '',
        selfClosing: true,
        raw: xml.slice(tagStart, gtIdx + 1),
      });
      i = gtIdx + 1;
      continue;
    }

    // Paired tag — find matching close, handling nesting of same tag
    const tagName = tagInner.split(/\s/)[0];
    const closeTag = '</' + tagName + '>';
    let depth = 1;
    let j = gtIdx + 1;

    while (j < xml.length && depth > 0) {
      const nextOpen = findNextOpenTag(xml, tagName, j);
      const nextClose = xml.indexOf(closeTag, j);

      if (nextClose === -1) {
        // Malformed — no matching close, skip this tag
        j = xml.length;
        break;
      }

      if (nextOpen !== -1 && nextOpen < nextClose) {
        depth++;
        j = nextOpen + 1;
      } else {
        depth--;
        if (depth === 0) {
          const closeEnd = nextClose + closeTag.length;
          children.push({
            tag: tagName,
            content: xml.slice(gtIdx + 1, nextClose),
            selfClosing: false,
            raw: xml.slice(tagStart, closeEnd),
          });
          i = closeEnd;
        } else {
          j = nextClose + 1;
        }
      }
    }

    if (depth > 0) {
      // Malformed paired tag, advance past the open tag
      i = gtIdx + 1;
    }
  }

  return children;
}

// Recursive AST evaluator
// counters = { nodeCount: 0, maxDepth: 0 } — mutated in place for final totals
function evalNode(xml, currentDepth = 0, counters = { nodeCount: 0, maxDepth: 0 }) {
  counters.nodeCount++;
  if (currentDepth > counters.maxDepth) counters.maxDepth = currentDepth;

  const trimmed = xml.trim();

  // --- Self-closing leaf: <leaf op="..." weight="..." valid="..."/> ---
  if (/^<leaf\s[^>]*\/>$/.test(trimmed)) {
    const attrs = parseLeafAttrs(trimmed);
    if (!attrs.op || !VALID_OPS.includes(attrs.op)) {
      return { op: 'ADD', weight: attrs.weight, valid: false, depth: currentDepth, reason: `unknown op: ${attrs.op}` };
    }
    return { op: attrs.op, weight: attrs.weight, valid: attrs.valid, depth: currentDepth };
  }

  // --- Determine outer tag ---
  const tagMatch = trimmed.match(/^<(\w+)[\s>]/);
  if (!tagMatch) {
    return { op: 'ADD', weight: 0, valid: false, depth: currentDepth, reason: 'malformed tag' };
  }
  const tagName = tagMatch[1].toLowerCase();

  // Get inner content between open and close tag
  const openTagEnd = trimmed.indexOf('>');
  const closeTag = '</' + tagName + '>';
  const closeTagIdx = trimmed.lastIndexOf(closeTag);
  if (openTagEnd === -1 || closeTagIdx === -1) {
    return { op: 'ADD', weight: 0, valid: false, depth: currentDepth, reason: `malformed ${tagName} tag` };
  }
  const inner = trimmed.slice(openTagEnd + 1, closeTagIdx);

  // --- <leaf op="..."></leaf> (paired, unusual but handle) ---
  if (tagName === 'leaf') {
    const attrs = parseLeafAttrs(trimmed);
    if (!attrs.op || !VALID_OPS.includes(attrs.op)) {
      return { op: 'ADD', weight: attrs.weight, valid: false, depth: currentDepth, reason: `unknown op: ${attrs.op}` };
    }
    return { op: attrs.op, weight: attrs.weight, valid: attrs.valid, depth: currentDepth };
  }

  // --- <route> — evaluate the single root child ---
  if (tagName === 'route') {
    const children = extractChildren(inner);
    if (children.length === 0) {
      return { op: 'ADD', weight: 0, valid: false, depth: currentDepth, reason: 'empty route' };
    }
    return evalNode(children[0].raw, currentDepth + 1, counters);
  }

  // --- <and> — ALL children must be valid; return highest-weight leaf ---
  if (tagName === 'and') {
    const children = extractChildren(inner);
    if (children.length === 0) {
      return { op: 'ADD', weight: 0, valid: false, depth: currentDepth, reason: 'and: no children' };
    }
    let best = null;
    for (const child of children) {
      const result = evalNode(child.raw, currentDepth + 1, counters);
      if (!result.valid) {
        return { op: result.op, weight: result.weight, valid: false, depth: currentDepth, reason: 'and: child invalid' };
      }
      if (!best || result.weight > best.weight) best = result;
    }
    return { ...best, depth: currentDepth };
  }

  // --- <or> — return the FIRST valid child ---
  if (tagName === 'or') {
    const children = extractChildren(inner);
    for (const child of children) {
      const result = evalNode(child.raw, currentDepth + 1, counters);
      if (result.valid) return { ...result, depth: currentDepth };
    }
    return { op: 'ADD', weight: 0, valid: false, depth: currentDepth, reason: 'or: no valid child' };
  }

  // --- <not> — negate validity of its single child ---
  if (tagName === 'not') {
    const children = extractChildren(inner);
    if (children.length === 0) {
      return { op: 'ADD', weight: 0, valid: false, depth: currentDepth, reason: 'not: no child' };
    }
    const result = evalNode(children[0].raw, currentDepth + 1, counters);
    return { ...result, valid: !result.valid, depth: currentDepth };
  }

  return { op: 'ADD', weight: 0, valid: false, depth: currentDepth, reason: `unknown tag: ${tagName}` };
}

// Entry point — strips to <route>...</route>, walks the AST
// Returns { op, weight, valid, ast_depth, node_count, reason? }
export function parseRoutingAST(xml) {
  const routeStart = xml.indexOf('<route');
  if (routeStart === -1) {
    // Allow a bare self-closing leaf as a degenerate case
    const leafTrimmed = xml.trim();
    if (/^<leaf\s[^>]*\/>$/.test(leafTrimmed)) {
      const counters = { nodeCount: 0, maxDepth: 0 };
      const result = evalNode(leafTrimmed, 0, counters);
      return { ...result, ast_depth: counters.maxDepth, node_count: counters.nodeCount };
    }
    return { op: 'ADD', weight: 0, valid: false, ast_depth: 0, node_count: 0, reason: 'no <route> element' };
  }

  // Verify boundary char after '<route' so we don't match '<router>'
  const charAfterRoute = xml[routeStart + 6]; // '<route' is 6 chars
  if (charAfterRoute !== '>' && charAfterRoute !== ' ' && charAfterRoute !== '\t' &&
      charAfterRoute !== '\n' && charAfterRoute !== '/' && charAfterRoute !== undefined) {
    return { op: 'ADD', weight: 0, valid: false, ast_depth: 0, node_count: 0, reason: 'no <route> element' };
  }

  const routeEnd = xml.lastIndexOf('</route>');
  if (routeEnd === -1) {
    return { op: 'ADD', weight: 0, valid: false, ast_depth: 0, node_count: 0, reason: 'unclosed <route>' };
  }
  const routeXml = xml.slice(routeStart, routeEnd + '</route>'.length);

  const counters = { nodeCount: 0, maxDepth: 0 };
  const result = evalNode(routeXml, 0, counters);
  return { ...result, ast_depth: counters.maxDepth, node_count: counters.nodeCount };
}

export { VALID_OPS, BOOLEAN_TAGS };

File: webllm/ast_parser.py
"""
ast_parser.py
XML boolean AST parser for sovereign routing expressions.
Direct port of ast_parser.mjs — no external XML library, pure regex recursion.
Handles: <route>, <and>, <or>, <not>, <leaf op="" weight="" valid=""/>
"""

import re
from typing import Optional

BOOLEAN_TAGS = frozenset(['and', 'or', 'not', 'route'])
VALID_OPS = ['ADD', 'MUL', 'XOR', 'LOOP', 'MEMCPY', 'MEMSET', 'STRCMP', 'HELLO']

_OP_RE     = re.compile(r'op="([^"]+)"')
_WEIGHT_RE = re.compile(r'weight="([^"]+)"')
_VALID_RE  = re.compile(r'valid="([^"]+)"')


def _parse_leaf_attrs(tag: str) -> dict:
    op_m     = _OP_RE.search(tag)
    weight_m = _WEIGHT_RE.search(tag)
    valid_m  = _VALID_RE.search(tag)
    return {
        'op':     op_m.group(1).upper() if op_m else None,
        'weight': float(weight_m.group(1)) if weight_m else 0.05,
        'valid':  valid_m.group(1) == 'true' if valid_m else False,
    }


def _find_next_open_tag(xml: str, tag_name: str, start: int) -> int:
    needle = '<' + tag_name
    i = start
    while i < len(xml):
        idx = xml.find(needle, i)
        if idx == -1:
            return -1
        char_after = xml[idx + 1 + len(tag_name)] if idx + 1 + len(tag_name) < len(xml) else None
        if char_after in ('>', ' ', '\t', '\n', '/', None):
            return idx
        i = idx + 1
    return -1


def _extract_children(xml: str) -> list:
    children = []
    i = 0
    while i < len(xml):
        while i < len(xml) and xml[i] != '<':
            i += 1
        if i >= len(xml):
            break
        if xml[i:i+4] == '<!--':
            end = xml.find('-->', i)
            i = end + 3 if end != -1 else len(xml)
            continue
        if xml[i+1:i+2] == '/':
            end = xml.find('>', i)
            i = end + 1 if end != -1 else len(xml)
            continue
        if xml[i+1:i+2] == '?':
            end = xml.find('?>', i)
            i = end + 2 if end != -1 else len(xml)
            continue

        tag_start = i
        gt_idx = xml.find('>', i)
        if gt_idx == -1:
            i += 1
            continue
        tag_inner = xml[i+1:gt_idx]

        # Self-closing
        if tag_inner.rstrip().endswith('/'):
            tag_name = re.split(r'[\s/]', tag_inner)[0]
            children.append({
                'tag': tag_name,
                'content': '',
                'self_closing': True,
                'raw': xml[tag_start:gt_idx+1],
            })
            i = gt_idx + 1
            continue

        # Paired tag
        tag_name = re.split(r'\s', tag_inner)[0]
        close_tag = '</' + tag_name + '>'
        depth = 1
        j = gt_idx + 1
        found = False
        while j < len(xml) and depth > 0:
            next_open  = _find_next_open_tag(xml, tag_name, j)
            next_close = xml.find(close_tag, j)
            if next_close == -1:
                j = len(xml)
                break
            if next_open != -1 and next_open < next_close:
                depth += 1
                j = next_open + 1
            else:
                depth -= 1
                if depth == 0:
                    close_end = next_close + len(close_tag)
                    children.append({
                        'tag': tag_name,
                        'content': xml[gt_idx+1:next_close],
                        'self_closing': False,
                        'raw': xml[tag_start:close_end],
                    })
                    i = close_end
                    found = True
                    break
                else:
                    j = next_close + 1
        if not found:
            i = gt_idx + 1
    return children


def _eval_node(xml: str, current_depth: int = 0, counters: Optional[dict] = None) -> dict:
    if counters is None:
        counters = {'node_count': 0, 'max_depth': 0}
    counters['node_count'] += 1
    counters['max_depth'] = max(counters['max_depth'], current_depth)

    trimmed = xml.strip()

    # Self-closing leaf
    if re.match(r'^<leaf\s[^>]*/>$', trimmed):
        attrs = _parse_leaf_attrs(trimmed)
        if not attrs['op'] or attrs['op'] not in VALID_OPS:
            return {'op': 'ADD', 'weight': attrs['weight'], 'valid': False,
                    'depth': current_depth, 'reason': f"unknown op: {attrs['op']}"}
        return {'op': attrs['op'], 'weight': attrs['weight'], 'valid': attrs['valid'],
                'depth': current_depth}

    tag_m = re.match(r'^<(\w+)[\s>]', trimmed)
    if not tag_m:
        return {'op': 'ADD', 'weight': 0, 'valid': False,
                'depth': current_depth, 'reason': 'malformed tag'}
    tag_name = tag_m.group(1).lower()

    open_tag_end = trimmed.find('>')
    close_tag = '</' + tag_name + '>'
    close_tag_idx = trimmed.rfind(close_tag)
    if open_tag_end == -1 or close_tag_idx == -1:
        return {'op': 'ADD', 'weight': 0, 'valid': False,
                'depth': current_depth, 'reason': f'malformed {tag_name} tag'}
    inner = trimmed[open_tag_end+1:close_tag_idx]

    if tag_name == 'leaf':
        attrs = _parse_leaf_attrs(trimmed)
        if not attrs['op'] or attrs['op'] not in VALID_OPS:
            return {'op': 'ADD', 'weight': attrs['weight'], 'valid': False,
                    'depth': current_depth, 'reason': f"unknown op: {attrs['op']}"}
        return {'op': attrs['op'], 'weight': attrs['weight'], 'valid': attrs['valid'],
                'depth': current_depth}

    if tag_name == 'route':
        children = _extract_children(inner)
        if not children:
            return {'op': 'ADD', 'weight': 0, 'valid': False,
                    'depth': current_depth, 'reason': 'empty route'}
        return _eval_node(children[0]['raw'], current_depth + 1, counters)

    if tag_name == 'and':
        children = _extract_children(inner)
        if not children:
            return {'op': 'ADD', 'weight': 0, 'valid': False,
                    'depth': current_depth, 'reason': 'and: no children'}
        best = None
        for child in children:
            result = _eval_node(child['raw'], current_depth + 1, counters)
            if not result['valid']:
                return {'op': result['op'], 'weight': result['weight'], 'valid': False,
                        'depth': current_depth, 'reason': 'and: child invalid'}
            if best is None or result['weight'] > best['weight']:
                best = result
        return {**best, 'depth': current_depth}

    if tag_name == 'or':
        children = _extract_children(inner)
        for child in children:
            result = _eval_node(child['raw'], current_depth + 1, counters)
            if result['valid']:
                return {**result, 'depth': current_depth}
        return {'op': 'ADD', 'weight': 0, 'valid': False,
                'depth': current_depth, 'reason': 'or: no valid child'}

    if tag_name == 'not':
        children = _extract_children(inner)
        if not children:
            return {'op': 'ADD', 'weight': 0, 'valid': False,
                    'depth': current_depth, 'reason': 'not: no child'}
        result = _eval_node(children[0]['raw'], current_depth + 1, counters)
        return {**result, 'valid': not result['valid'], 'depth': current_depth}

    return {'op': 'ADD', 'weight': 0, 'valid': False,
            'depth': current_depth, 'reason': f'unknown tag: {tag_name}'}


def parse_routing_ast(xml: str) -> dict:
    route_start = xml.find('<route')
    if route_start == -1:
        leaf_trimmed = xml.strip()
        if re.match(r'^<leaf\s[^>]*/>$', leaf_trimmed):
            counters = {'node_count': 0, 'max_depth': 0}
            result = _eval_node(leaf_trimmed, 0, counters)
            return {**result, 'ast_depth': counters['max_depth'],
                    'node_count': counters['node_count']}
        return {'op': 'ADD', 'weight': 0, 'valid': False,
                'ast_depth': 0, 'node_count': 0, 'reason': 'no <route> element'}

    char_after = xml[route_start + 6] if route_start + 6 < len(xml) else None
    if char_after not in ('>', ' ', '\t', '\n', '/', None):
        return {'op': 'ADD', 'weight': 0, 'valid': False,
                'ast_depth': 0, 'node_count': 0, 'reason': 'no <route> element'}

    route_end = xml.rfind('</route>')
    if route_end == -1:
        return {'op': 'ADD', 'weight': 0, 'valid': False,
                'ast_depth': 0, 'node_count': 0, 'reason': 'unclosed <route>'}

    route_xml = xml[route_start:route_end + len('</route>')]
    counters = {'node_count': 0, 'max_depth': 0}
    result = _eval_node(route_xml, 0, counters)
    return {**result, 'ast_depth': counters['max_depth'],
            'node_count': counters['node_count']}
